lttng-kallsyms.c

```diff
@@ -0,0 +1,63 @@
+/* SPDX-License-Identifier: (GPL-2.0 or LGPL-2.1)
+ *
+ * lttng-kallsyms.c
+ *
+ * Copyright (C) 2019 Geneviève Bastien <gbastien@versatic.net>
+ */
+
+#include <linux/module.h>
+#include <linux/kallsyms.h>
+#include <lttng-events.h>
+#include <lttng-kallsyms.h>
+
+/* Define the tracepoints, but do not build the probes */
+#define CREATE_TRACE_POINTS
+#define TRACE_INCLUDE_PATH instrumentation/events/lttng-module
+#define TRACE_INCLUDE_FILE lttng-kallsyms
+#define LTTNG_INSTRUMENTATION
+#include <instrumentation/events/lttng-module/lttng-kallsyms.h>
+
+DEFINE_TRACE(lttng_kallsyms_kernel_symbol);
+
+/*
+ * Trace the kernel symbols from a given module
+ *
+ * data: The lttng_session instance
+ * symbol_name: The function name this symbol resolves to
+ * module: The module containing this symbol. Can be NULL
+ * symbol_addr: The symbol address
+ */
+static
+int _lttng_one_symbol_received(void * data, const char * symbol_name,
+		struct module * module, unsigned long symbol_addr)
+{
+	struct lttng_session *session = data;
+
+	if (module) {
+		trace_lttng_kallsyms_kernel_symbol(session, symbol_addr,
+				symbol_name, module->name);
+	} else {
+		trace_lttng_kallsyms_kernel_symbol(session, symbol_addr,
+				symbol_name, "");
+	}
+
+	return 0;
+}
+
+int lttng_enumerate_kernel_symbols(struct lttng_session *session)
+{
+	int ret = 0;
+
+	ret = kallsyms_on_each_symbol(_lttng_one_symbol_received, (void *) session);
```

Comment:
AFAIU, the modules_mutex must be held around this call.

module_assert_mutex() called by module_kallsyms_on_each_symbol() should have caught this. Did you build and test with a kernel configuration including additional kernel debugging ?

Reply:
I thought it was being taken care of by the caller/callee? Is that what you mean by caught this?

As for building and testing, I did a few kernel compilation, namely with/without KALLSYMS_ALL. Tried without KALLSYMS, but then the whole lttng-modules doesn't work. As you said, it's a prerequisite. What other kernel options should be tested? I'm setup to compile anything you wish now!

Reply:
see the Linux kernel Documentation/process/submit-checklist.rst. It includes the various configuration options under which new features need to be tested. This also applies to lttng-modules.

By "caught this", I mean that this assertion should trigger a dmesg warning. But it will only be effective when the kernel is built with the right debug config options.

What caller/callee are you referring to exactly ? You are implementing the caller, so it is up to you to hold the appropriate locks when calling this kernel API. Unless you can point to a specific code line which is responsible for internally taking those locks.

Reply:
oh! I misunderstood, I though module_assert_mutex() was actively taking the mutex, not just checking I had it, my bad